bpo-41176: revise Tkinter mainloop dispatching flag behavior

richardsheridan opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-41176](https://bugs.python.org/issue41176) -->
https://bugs.python.org/issue41176
<!-- /issue-number -->

richardsheridan committed:
bpo-41176: Expose tkinter dispatching state

new method `dispatching` on tkinter, Misc, and _tkinter.tkapp

richardsheridan committed:
bpo-41176: Deprecate implicit mainloop waiting behavior

DeprecationWarning issued on implicit wait
Docstring changes
Comment stubs for eventual removal of WaitForMainloop
Add setmainloopwaitattempts method to _tkinter.tkapp

richardsheridan committed:
bpo-41176: additional docstring corrections

richardsheridan committed:
bpo-41176: Correct tkinter test usage of `update`

Replace all usages of update_idletasks in tests with update.

update_idletasks does not in fact process all pending events.
It updates the idle tasks only, without touching the normal event queue. This is not normaly a problem... until it is.

Update docstrings to clarify.

E-Paine commented on Lib/tkinter/test/test_tkinter/test_misc.py:
```diff
@@ -192,6 +192,51 @@ def test_clipboard_astral(self):
+            self.assertFalse(thing.dispatching())
+
+    def test_thread_must_wait_for_mainloop(self):
+        import threading, time
```
If there is a reason the imports are here, please add a comment to say so to stop people like me moaning about PEP 8 ;-)

E-Paine commented on Lib/tkinter/test/test_tkinter/test_misc.py:
```diff
@@ -192,6 +192,51 @@ def test_clipboard_astral(self):
+            #mainloop, dispatching, willdispatch
+            self.assertFalse(thing.dispatching())
+            root.after(0, lambda:self.assertTrue(thing.dispatching()))
+            root.after(0, root.quit)
```
We **need** to guarantee the quit method is called after the mainloop call, otherwise the test will hang indefinitely

richardsheridan replied:
Tcl provides as close to a guarantee of this as you can get with the `after` command.  Do you mean that we should make sure the test doesn't hang in the face of future bugs?

E-Paine replied:
Running `root.after(0, print, "tst")` in the console immediately prints "tst" even though the mainloop hasn't been called. I haven't investigated the cause but it seems it doesn't necessarily wait for the mainloop to be called.

richardsheridan replied:
That is an artifact of the REPL EventHook. It is [deep magic](https://stackoverflow.com/questions/20891710/what-magic-prevents-tkinter-programs-from-blocking-in-interactive-shell). It should not be an issue in a script.

richardsheridan replied:
This brings a question to mind, should EventHook set the dispatching flag as well? We should at least document a clear cut distinction between REPL behavior and script behavior somewhere.

E-Paine replied:
I believe it would make sense to set dispatching in EventHook. Internally, I don't think it makes a difference but for users who call the `dispatching` method it could be quite useful (so long as we properly document this behaviour to avoid confusion). And yes, we should definitely document the REPL behaviour (though that should probably be a different issue & PR)!

richardsheridan replied:
Then I propose we leave the return value of `dispatching` as PyLong rather than PyBool. For this PR it will be a one-line change, but for the future we can its value depending on what is dispatching and map that to an IntEnum for REPL, mainloop, ipython, and whoever else.

E-Paine commented on Lib/tkinter/test/test_tkinter/test_misc.py:
```diff
@@ -192,6 +192,51 @@ def test_clipboard_astral(self):
+    def test_thread_must_wait_for_mainloop(self):
+        import threading, time
+        def target():
+            self.assertFalse(root.dispatching())
```
An assertion in a thread will not cause the test to fail, and it instead needs to be raised in the main thread. I would propose using a `nonlocal` variable to catch errors in the thread to reraise them after the `thread.join()` call

E-Paine commented on Lib/tkinter/test/test_tkinter/test_misc.py:
```diff
@@ -192,6 +192,51 @@ def test_clipboard_astral(self):
+
+        ready_for_mainloop = threading.Event()
+        thread = threading.Thread(target=target)
+        self.assertFalse(root.dispatching())
```
Do we need the same assertion in the thread and out?

richardsheridan replied:
Not if assertions in the thread are ignored!

richardsheridan replied:
To actually answer the question, yes. If the flag changes between the main thread and the target thread at this point, I want to know about it.

E-Paine commented on Lib/tkinter/test/test_tkinter/test_misc.py:
```diff
@@ -192,6 +192,51 @@ def test_clipboard_astral(self):
+            # self.assertTrue(root.dispatching()) but patient
+            t = time.monotonic()
+            while not root.dispatching():
+                time.sleep(.001)
```
Is this too small and could cause the main thread to hang?

E-Paine commented on Modules/_tkinter.c:
```diff
@@ -360,21 +360,36 @@ Sleep(int milli)
+                "It seems you are implicitly waiting for "
+                "the mainloop to come up.\n"
+                "This behavior is deprecated; consider polling "
+                "dispatching() instead.\n", 1))
```
Is this warning better suited to just before the for loop (after the if)? That way the warning is raised immediately when waiting and you can have both the warning and RuntimeError

richardsheridan replied:
That is a happy, non-deprecated path when using `_tkinter.tkapp.setmainloopwaitattempts(0)`. This function should warn on timeout too though. I've added a fix by duplicating code, but if you have a better idea, I'm open to it.

E-Paine replied:
True, I forgot about `mainloopwaitattempts` being equal to 0. To avoid duplicate code, I would put it where I originally suggested, but just add a check to ensure it is only raised when `mainloopwaitattempts` is not 0.

richardsheridan committed:
bpo-41176: Rename according to comment and use Clinic

E-Paine commented on Modules/_tkinter.c:
```diff
@@ -3022,17 +3048,86 @@ Tkapp_WantObjects(PyObject *self, PyObject *args)
+_tkinter_tkapp_dispatching_impl(TkappObject *self)
+/*[clinic end generated code: output=1b0192766b008005 input=fca22ac098f764ff]*/
+{
+    return PyLong_FromLong(self->dispatching);
```
```suggestion
    return PyBool_FromLong(self->dispatching);
```
I think we are better returning a bool than int

E-Paine commented on Modules/clinic/_tkinter.c.h:
```diff
@@ -631,7 +631,14 @@ _tkinter_tkapp_loadtk(TkappObject *self, PyObject *Py_UNUSED(ignored))
+"\n"
+"DEPRECATED\n"
+"\n"
+"call(), var_invoke(), and createcommand() will wait indefinitely for main loop\n"
```
"indefinitely"? Sorry if I'm going crazy, but is the wait not due to a call to WaitForMainloop (in which case, not indefinitely)?

richardsheridan replied:
You aren't crazy, this code is crazy!

Willdispatch fakes out waitformainloop so the thread ends up sitting on a Tcl_conditionwait in tkapp_threadsend in the hopes that you "willdispatch" at some point in the future, and put the thread send event through the queue in the main thread with update or dooneevent or mainloop.

E-Paine commented on Modules/clinic/_tkinter.c.h:
```diff
@@ -645,6 +652,61 @@ _tkinter_tkapp_willdispatch(TkappObject *self, PyObject *Py_UNUSED(ignored))
+"These are used for that call(), var_invoke(), and createcommand().\n"
+"\n"
+"Current default is 10 for a 1 second wait, but future behavior\n"
+"will be equivalent 0.\n"
```
```suggestion
"will be equivalent to 0.\n"
```
Sorry for being so picky!

E-Paine reviewed (commented):
This is generally looking very good, though there are a few things I feel need addressing:

1) Please sign your CLA! - Unfortunately, we cannot legally accept the patch until the CLA has been signed. Once it has been approved, an asterisk will appear next to your name on bugs.python and the bots should change the PR labels automatically.
2) New methods = new docs - As we are adding new methods, these should be fully documented in the docs (probably in Doc/library/tkinter.rst)
3) Whatsnew? - I feel this is definitely a big enough change to be worth adding to the 3.10 whatsnew
4) ACKS - Because this is your first non-trivial patch, please feel free to add yourself to Misc/ACKS

This is just an initial look at the code and I have not tested the patch yet, though I hope to do this soon.

Thank you for this PR: as I have already said it looks very good and the code has now persuaded me in favour of deprecating WaitForMainloop. I may, at some point, drag Terry into this for his opinion on the changes (and how they could affect IDLE)

richardsheridan committed:
bpo-41176: Address PR comments, add willdispatch test

Exceptions raised in the thread will be ignored, so capture the conditions of interest and assert them after joining.

Add WaitForMainloop warning on timeout too.

richardsheridan committed:
bpo-41176: deduplicate warning code

richardsheridan committed:
bpo-41176: rectify typo

richardsheridan committed:
bpo-41176: ACKS

richardsheridan committed:
bpo-41176: fix future behavior

Still need to raise if not dispatching

richardsheridan committed:
bpo-41176: inline thread flag checks

Too hard to tell which flag broke if using the loop

richardsheridan committed:
bpo-41176: internal flag should flip, not just be true

richardsheridan committed:
bpo-41176: improve dispatching docstring

richardsheridan committed:
bpo-41176: Update whatsnew

richardsheridan committed:
bpo-41176: whitespace

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

richardsheridan committed:
bpo-41176: data to method

richardsheridan committed:
bpo-41176: fix future behavior

brackets

richardsheridan committed:
bpo-41176: Make dispatching return an int

In preparation for supporting multiple dispatchers with an IntEnum

richardsheridan committed:
bpo-41176: attempt to expose CI error

failing with timeout gives no helpful debugging info, so try to eliminate all possible hangs

richardsheridan committed:
bpo-41176: revert internal memory of previous dispatching

Seems to cause CI to never exit mainloop at least on Ubuntu. No idea why yet.

richardsheridan committed:
bpo-41176: Attempt Ubuntu CI fix

Co-Authored-By: E-Paine <e-paine@users.noreply.github.com>

richardsheridan committed:
bpo-41176: Attempt Ubuntu CI fix

quit may not quit if no events are forthcoming, but after creates its own event to dispatch while calling quit.

Co-Authored-By: E-Paine <e-paine@users.noreply.github.com>

E-Paine commented on Lib/tkinter/__init__.py:
```diff
@@ -1437,7 +1437,12 @@ def dispatching(self):
-        """Quit the Tcl interpreter. All widgets will be destroyed."""
+        """Signal the Tkinter mainloop to stop dispatching.
+
+        The mainloop will exit AFTER THE NEXT Tcl event handler is called.
```
I am not sure I agree. Looking at the source, there should be no reason that Tcl should need to be called, which is why the problem is so confusing.
```suggestion
        The mainloop will exit AFTER THE CURRENT Tcl event handler is finished calling.
```
(you will probably need to fix the line wrapping)

richardsheridan replied:
Hm, the reason I said NEXT is that it is possible to call quit without incurring an event (our bug). So CURRENT would be at best undefined in that case.  However, CURRENT would absolutely be the correct way to describe what is happening in after(0, quit) or if any callback, command, or binding calls quit.  Is it too verbose to say "AFTER the current Tcl event handler is finished calling, or the next if quit is called outside the context of an event."?

E-Paine replied:
The reason I said next was that `quit` is not a Tcl command. Instead is sets the `quitMainloop` variable to `1`, though the chance of changing the value between its [assignment to `0` and the while loop](https://github.com/python/cpython/blob/master/Modules/_tkinter.c#L2869-L2873) is *incredibly* small (sorry, I didn't quite understand the point you were trying to make so I am assuming it was this). If `quitMainloop` is set to `1` during one iteration (one Tcl update), the while loop stops before the next iteration.

richardsheridan committed:
bpo-41176: Quit, mainloop, dispatch docstrings

richardsheridan committed:
bpo-41176: WIP Docs in tkinter.rst

richardsheridan committed:
bpo-41176: satisfy Doc make check

richardsheridan committed:
bpo-41176: use local max attempts variable

prevent subsequent calls to setmainloopwaitattempts from stepping on the behavior of current WaitForMainloop calls.

richardsheridan committed:
bpo-41176: dispatching returns PyBool

richardsheridan committed:
bpo-41176: update docstrings

richardsheridan committed:
bpo-41176: remove old deprecation warning

richardsheridan committed:
bpo-41176: update test with new deprecation and new future behavior

richardsheridan committed:
bpo-41176: update Docs

richardsheridan committed:
bpo-41176: Comment new post-deprecation plan

richardsheridan committed:
bpo-41176: New behavior plan: Doc and Warn fixes

E-Paine commented on Doc/library/tkinter.rst:
```diff
@@ -95,6 +88,29 @@ Or, more often::
 
    .. FIXME: The following keyword arguments are currently recognized:
 
+.. method:: Tk.mainloop(threshold)
```
Maybe just add a small thing above saying "here is a brief description of a few of the methods provided by the tk class..." (or something of the like)

E-Paine commented on Doc/library/tkinter.rst:
```diff
@@ -95,6 +88,29 @@ Or, more often::
+   Determines if the Tkinter main loop is running. Returns True if the main loop is running, or
+   returns False if the main loop is not running. It is possible for some entity other than
+   the main loop to be dispatching events. Some examples are: calling the :meth:`Tk.update` command,
+   :meth:`_tkinter.tkapp.doonevent`, and the python command line EventHook.
```
Let people know this method is new using `.. versionadded:: 3.10` at the end of the description

E-Paine commented on Doc/library/tkinter.rst:
```diff
@@ -106,6 +122,58 @@ Or, more often::
    created by the :func:`Tcl` object can have a Toplevel window created (and the Tk
    subsystem initialized) by calling its :meth:`loadtk` method.
 
+.. TODO: Not a class, a module. how do I inline another module into this RST file?
```
Sorry for being a pain and asking for this: I don't think it is suited to this file (people can use the `help` method if they want). While I don't think it is suited here, *please* don't lose the work that has gone into it, and instead keep a local copy of the following for if the future tkinter internals doc goes ahead.

E-Paine commented on Doc/whatsnew/3.10.rst:
```diff
@@ -117,6 +117,22 @@ Add :data:`sys.orig_argv` attribute: the list of the original command line
 (Contributed by Victor Stinner in :issue:`23427`.)
 
+tkinter
+---
```
It is general practice for the title to be completely underlined by dashes (in this case, 4 more dashes are required)

E-Paine commented on Doc/whatsnew/3.10.rst:
```diff
@@ -117,6 +117,22 @@ Add :data:`sys.orig_argv` attribute: the list of the original command line
+Added another new method :meth:`_tkinter.tkapp.setmainloopwaitattempts` which
+can be used to obtain future thread waiting behavior.
+
+(see also Deprications section)
```
```suggestion
(see also Deprecations section)
```

E-Paine commented on Doc/whatsnew/3.10.rst:
```diff
@@ -117,6 +117,22 @@ Add :data:`sys.orig_argv` attribute: the list of the original command line
+tkinter
+---
+
+Added :meth:`tkinter.dispatching`, :meth:`tkinter.Misc.dispatching`, and
```
The following description is quite long, so I would XXX it to bring it to the attention of the moderator.

E-Paine commented on Doc/whatsnew/3.10.rst:
```diff
@@ -130,6 +146,13 @@ Optimizations
 Deprecated
 ==========
 
+The :mod:`tkinter` module has deprected the undocumented behavior where a
```
Again, probably worth "XXX"ing

E-Paine commented on Lib/tkinter/test/test_tkinter/test_misc.py:
```diff
@@ -1,5 +1,7 @@
 import unittest
 import tkinter
+import threading, time
```
```suggestion
import threading
```

E-Paine commented on Lib/tkinter/test/test_tkinter/test_misc.py:
```diff
@@ -192,6 +194,100 @@ def test_clipboard_astral(self):
+        root = self.root
+        self.assertFalse(root.dispatching())
+        with self.assertWarns(DeprecationWarning):
+            root.tk.willdispatch()
```
Can we reset `dispatching` to 0 so as to not interfere with future tests? (this is not currently causing issues because tests are sorted alphabetically and so this is the last of the new tests to be run - we cannot rely on them being sorted)

richardsheridan replied:
I think that root is destroyed and re-created each test in the `AbstractTkTest` `tearDownClass` and `setUpClass` methods respectively. Also presently the only way to programatically reset the dispatching flag at this moment is to call `mainloop`

E-Paine replied:
I just added a `print` statement to `setUpClass` and can confirm it is only called once per test class.

> the only way to programatically reset the dispatching flag at this moment is to call `mainloop`

Then that's what we'll have to do (using the old `root.after(0, root.quit)`)

richardsheridan replied:
> I just added a `print` statement to `setUpClass` and can confirm it is only called once per test class.

wow TIL

E-Paine commented on Lib/tkinter/test/test_tkinter/test_misc.py:
```diff
@@ -192,6 +194,100 @@ def test_clipboard_astral(self):
+
+            # Everything after the event set must go in this try
+            # to guarantee that root.quit() is called in finally
+            try:
```
Now we have established the source of the timeout issue, is the `try` still required?

richardsheridan replied:
Not really, crashes in `dispatching`, `range`, or `sleep` should be exceedingly rare.

E-Paine commented on Lib/tkinter/test/test_tkinter/test_misc.py:
```diff
@@ -192,6 +194,100 @@ def test_clipboard_astral(self):
+
+        try:
+            ready_for_mainloop = threading.Event()
+            thread = threading.Thread(target=target, daemon=True)
```
Does it need to be daemon?

E-Paine commented on Modules/_tkinter.c:
```diff
@@ -360,21 +360,35 @@ Sleep(int milli)
-    for (i = 0; i < 10; i++) {
-        if (self->dispatching)
-            return 1;
+    int i = 0;
```
```suggestion
    int i;
```

E-Paine commented on Modules/_tkinter.c:
```diff
@@ -360,21 +360,35 @@ Sleep(int milli)
+        "limit with setmainloopwaitattempts().\n"
+        "If you want to avoid this wait consider polling "
+        "dispatching() before issuing commands from a thread.\n", 1
+    );
```
What's the reasoning for this warning being here?

richardsheridan replied:
I think it makes more sense to be here considering my recent change to the planned future behavior.  Hitting this error after one second is the actual deprecated behavior, rather than waiting at all.

E-Paine commented on Modules/_tkinter.c:
```diff
@@ -3022,17 +3040,89 @@ Tkapp_WantObjects(PyObject *self, PyObject *args)
+/*[clinic end generated code: output=0e3f46d244642155 input=7b46376275304a4a]*/
 {
+    if (PyErr_WarnEx(PyExc_DeprecationWarning,
+        "willdispatch() is deprecated; in the future it will have no effect.", 1))
```
So its not being removed? Not saying this is bad at all, but I would appreciate hearing your reasoning.

richardsheridan replied:
Right, with the future behavior of waiting indefinitely, the only change necessary for deprecation will be to keep this method from corrupting the return value of `dispatching`. So we could remove it or not, but leaving it as a null op would break less code. I suppose it's a point we should bring up with Serhiy and/or Guilherme when the time comes.

E-Paine commented on Modules/clinic/_tkinter.c.h:
```diff
@@ -645,6 +652,65 @@ _tkinter_tkapp_willdispatch(TkappObject *self, PyObject *Py_UNUSED(ignored))
+    PyObject *return_value = NULL;
+    int new_val;
+
+    new_val = _PyLong_AsInt(arg);
```
```suggestion
    int new_val = _PyLong_AsInt(arg);
```
(removing previous `int new_val`)

richardsheridan replied:
That bit was created by clinic, is it ok to mess with it?

E-Paine replied:
If it was created by clinic, that's fine (just wanted to make it a little neater if possible)

richardsheridan committed:
bpo-41176: Resolve review issues

richardsheridan committed:
bpo-41176: Title underlines should be same length as title

richardsheridan committed:
bpo-41176:  reset dispatching flag in test

richardsheridan committed:
bpo-41176: layout Tk methods

richardsheridan committed:
bpo-41176: clarify dispatching() limitations

tacaswell commented on Lib/tkinter/__init__.py:
```diff
@@ -590,9 +590,20 @@ def get(self):
-def mainloop(n=0):
-    """Run the main loop of Tcl."""
-    _default_root.tk.mainloop(n)
+def mainloop(threshold=0):
```
This is an API change that should be flagged?

richardsheridan replied:
Good point, this was just meant to make the variable name a bit more descriptive, but it could break (rare) user code that relied on setting `n` as a keyword argument.
